posts/2023-12-21-async-fn-rpit-in-traits.md

```diff
@@ -0,0 +1,229 @@
+---
+layout: post
+title: "Announcing `async fn` and return-position `impl Trait` in traits"
+author: Tyler Mandry
+team: The Async Working Group <https://www.rust-lang.org/governance/wgs/wg-async>
+---
+
+The Rust Async Working Group is excited to announce major progress towards our goal of enabling the use of `async fn` in traits. The next Rust release will include stable support for both `-> impl Trait` notation and `async fn` in traits.
+
+This is a big milestone, and we know many users will be itching to try these out in their own code. However, we are still missing some important features that many users need. Read on for an recommendations on when and how to use the stabilized features.
+
+## What's stabilizing
+
+Ever since [RFC #1522], Rust has allowed users to write `impl Trait` as the return type of functions. This means that the function returns "some type that implements `Trait`". This is commonly used to return iterators, closures, or other types that are complex or even impossible to write explicitly:
```

Comment:
```suggestion
Ever since the stabilization of [RFC #1522] in Rust 1.26, Rust has allowed users to write `impl Trait` as the return type of functions (often called "RPIT"). This means that the function returns "some type that implements `Trait`". This is the only way to return types that cannot be named, such as futures and closures, without boxing. It's also used when an API author wants to return a type with a well defined interface, such as an iterator, without making a stability commitment to the exact type (such as, e.g., `Zip<Once<T>, Once<T>>`).
```

The changes here:

- It may be better to note the release in which RPIT stabilized, as in some sense that's when it became available.
- It's probably good to mention "RPIT" here somewhere, as many people do know and may scan for that term, and it helps connect this to "RPITIT" below.
- The framing of RPIT as being for "complex" types has proved to be misleading (this language originates in the RFC).  The choice of whether to return an opaque type (when the type *could* instead be returned concretely) hinges on whether you want that type to be *opaque*, with all that implies for the limitations of how it can be used by the caller and what it means for the stability commitment of the callee.  We don't want to suggest to people that they should switch to RPIT just because the type has become "complex" (plenty of complex type should be returned concretely).  So it may be better to frame the choice to use RPIT as being for when 1) the type cannot be named or when 2) we want to not make a SemVer commitment to the concrete type.

Reply:
I would say using RPIT for complex types is a common and acceptable practice – just not in public APIs.